File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX - 1);
        dp[0] = 0;
        for(int coin: coins)
            if(coin <= amount)
                dp[coin] = 1;
        int diff;
        for(int i = 1; i <= amount; ++i) {
            for(int coin: coins) {
                diff = i - coin;
                if(diff >= 0)
                    dp[i] = min(dp[i], 1 + dp[diff]);
            }
        }
        return (dp[amount] >= INT_MAX - 1) ? -1: dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.java
public class Solution {

  public int coinChange(int[] coins, int amount) {
    return coinChange(coins, amount, new int[amount]);
  }

  private int coinChange(int[] coins, int rem, int[] count) {
    if (rem == 0) return 0;
    if (count[rem - 1] != 0) return count[rem - 1];
    int min = Integer.MAX_VALUE;
      int diff;
    for (int coin : coins) {
        diff = rem - coin;
        if (diff < 0)
            continue;
      int res = coinChange(coins, diff, count);
      if (res >= 0 && res < min)
        min = 1 + res;
    }
    count[rem - 1] = (min == Integer.MAX_VALUE) ? -1 : min;
    return count[rem - 1];
  }
}